packages/dds/merge-tree/src/attributionCollection.ts

```diff
@@ -5,32 +5,66 @@
 
 import { assert } from "@fluidframework/common-utils";
 import { RedBlackTree } from "./collections";
-import { compareNumbers, ISegment as ISegmentCurrent } from "./mergeTreeNodes";
-
-// TODO: Once integrated into merge-tree, this interface can be removed
-interface ISegment extends ISegmentCurrent {
-    attribution?: AttributionCollection<unknown>;
-}
+import { AttributionKey, compareNumbers, ISegment } from "./mergeTreeNodes";
 
 export interface SerializedAttributionCollection {
-    keys: unknown[];
+    /**
+     * Parallel array with posBreakpoints which tracks the seq of insertion.
+     * Ex: if seqs is [45, 46] and posBreakpoints is [0, 3], the section of the string
+     * between offsets 0 and 3 was inserted at seq 45 and the section of the string between
+     * 3 and the length of the string was inserted at seq 46.
+     */
+    seqs: number[];
     posBreakpoints: number[];
     /* Total length; only necessary for validation */
     length: number;
 }
 
-export class AttributionCollection<T> {
-    private readonly entries: RedBlackTree<number, T> = new RedBlackTree(compareNumbers);
+/**
+ * @alpha
+ */
+export interface IAttributionCollection<T> {
+    /**
+     * Retrieves the attribution key associated with the provided offset.
+     */
+    getAtOffset(offset: number): T;
+
+    /**
+     * Total length of all attribution keys in this collection.
+     */
+    readonly length: number;
 
-    public constructor(baseEntry: T, private _length: number) {
+    /**
+     * Retrieve all key/offset pairs stored on this segment. Entries should be ordered by offset, such that
+     * the `i`th result's attribution key applies to offsets in the open range between the `i`th offset and the
+     * `i+1`th offset.
+     * The last entry's key applies to the open interval from the last entry's offset to this collection's length.
+     * @internal
+     */
+    getAll(): Iterable<{ offset: number; key: T; }>;
+
+    /** @internal */
+    splitAt(pos: number): IAttributionCollection<T>;
+
+    /** @internal */
+    append(other: IAttributionCollection<T>): void;
+
+    /** @internal */
+    clone(): IAttributionCollection<T>;
+}
+
+export class AttributionCollection implements IAttributionCollection<AttributionKey> {
```

Comment:
its a bit odd the interface is generic, but the class is not

Reply:
yeah, this is because it's internally storing just `seq` but the API will eventually want to return `AttributionKey`. Rather than try and introduce weird abstractions I just hardcoded the "add `{ type: "op" }` aspect, which makes the implementation non-generic. I suspect this will be cleaned up somewhat with better merge-tree extensibility points. It's not exposed to the public API in any case, so shouldn't be a big deal.